src/ORM/Table.php

```diff
@@ -530,10 +534,46 @@ public function setSchema($schema)
         }
 
         $this->_schema = $schema;
+        if (Configure::read('debug')) {
+            $this->checkAliasLengths();
+        }
 
         return $this;
     }
 
+    /**
+     * Checks if all table name + column name combinations used for
+     * queries fit into the max length allowed by database driver.
+     *
+     * @return void
+     * @throws \RuntimeException When an alias combination is too long
+     */
+    protected function checkAliasLengths(): void
+    {
+        if ($this->_schema === null) {
+            throw new RuntimeException("Unable to check max alias lengths for  `{$this->getAlias()}` without schema.");
+        }
+
+        $maxLength = null;
+        if (method_exists($this->getConnection()->getDriver(), "getMaxAliasLength")) {
+            $maxLength = $this->getConnection()->getDriver()->getMaxAliasLength();
+        }
+        if ($maxLength === null) {
+            return;
+        }
+
+        $table = $this->getAlias();
+        foreach ($this->_schema->columns() as $name) {
+            if (strlen($table . '__' . $name) > $maxLength) {
+                $nameLength = $maxLength - 2;
+                throw new RuntimeException(
+                    "The combination of `{$table}` and `{$name}` must be less than {$nameLength}. " .
+                    "You must shorten the one of those names before `Table` can generate queries."
```

Comment:
This will be a hard failure for some folks, and they won't be able to use the ORM without schema changes. While I think the early failure is good we might want to suggest that they alter the column name, or rename the table to be more explicit on what the next step should be.

Reply:
By hard failure you mean their ORM is broken and they don't realize why? I put it in Table because any query you make at the ORM level needs this to work, right?